src/Widgets/PowerModeList.vala

```diff
@@ -0,0 +1,170 @@
+/*
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ * SPDX-FileCopyrightText: 2024 elementary, Inc. (https://elementary.io)
+ */
+
+using Power.Services.DBusInterfaces;
+
+public class Power.Widgets.PowerModeList : Gtk.Box {
+    public static bool successfully_initialized { get; private set; default = true; }
+
+    private static PowerProfile? pprofile;
+    private static GLib.Settings? settings;
+
+    public string settings_key { get; construct; }
+
+    private Gtk.RadioButton saver_radio;
+    private Gtk.RadioButton balanced_radio;
+    private Gtk.RadioButton performance_radio;
+
+    private const string ICON_RES = "/io/elementary/desktop/wingpanel/power/scalable/categories/";
+
+    public PowerModeList (bool on_battery) {
+        Object (
+            settings_key: on_battery ? "profile-on-good-battery" : "profile-plugged-in",
+            orientation: Gtk.Orientation.VERTICAL,
+            spacing: 0,
+            margin: 6
+        );
+    }
+
+    static construct {
+        var schema = SettingsSchemaSource.get_default ().lookup ("io.elementary.settings-daemon.power", true);
+        if (schema != null && schema.has_key ("profile-plugged-in") && schema.has_key ("profile-on-good-battery")) {
+            settings = new GLib.Settings ("io.elementary.settings-daemon.power");
+        } else {
+            warning ("settings-daemon schema not found, will connect to power-profiles-daemon directly");
+        }
+
+        try {
+            pprofile = Bus.get_proxy_sync (BusType.SYSTEM, POWER_PROFILES_DAEMON_NAME, POWER_PROFILES_DAEMON_PATH, DBusProxyFlags.NONE);
+        } catch (Error e) {
+            critical (e.message);
+            successfully_initialized = false;
+        }
+    }
+
+    construct {
+        if (pprofile == null) {
+            return;
+        }
+
+        build_ui ();
```

Comment:
`construct` is already your kind of `build_ui` function. it's the function that gets called once when the class is initialized, so we don't need to make a separate `build_ui` function

Reply:
I usually separate the ui building code and the event handling code in two separate functions. I can put it all in one place if you say, it's just my style.

Reply:
I'm not gonna block on it, but we usually don't do this